### ISSUE TITLE
Hide tag weights and change negative weight tags to "Hidden tags"

### ISSUE THREAD

**Issue by PhilipHow** [forem/forem#19675]
## Is this feature related to a problem? Please describe.

- As a feed-consuming user, I want a simple means of hiding content I do not want to see in the feed. For example, if I am uninterested in javascript content, I can hide the `#javascript` tag.

## Describe the solution you’d like

This work targets two overall goals:

- Increase feed engagement and perceived feed quality by giving more control to users over what appears in their feed
- Improve the UX of the "Following tags" area in the user's dashboard by removing tag weights.

To reduce the complexity of this work, we will **continue to use the existing tag weights under the hood** and use the existing impact of negative tag weights on the feed (and nowhere new in the product, for the time being). In the future, we may split this work into its own logic and may apply it to other areas (e.g., notifications).

Feed modifications (using the existing feed variant mechanism) will happen in parallel with this work **but are NOT part of this issue**. It is imagined that the feed will now only interact with tag weights in a binary fashion (positive or negative) as opposed to using the tag-weighted value.

**This issue should not touch the feed algorithm or alter any feed math/calculations as this work should only touch the UI. QA may result in changing the tag weight to a <-1.** 

### Designs

Here is  the [Figma link](https://www.figma.com/file/nj32OuyDOR430frESjcOCM/Tag-Interaction?type=design&node-id=1637-1693&mode=design&t=1a924zsOXwQZJNSN-4) and below are small interactions to Hide/Unhide tag

![Tag page](https://github.com/forem/forem/assets/84626054/7de36496-ebab-47bb-ac1e-b5037ae97272)

![Following tags page](https://github.com/forem/forem/assets/84626054/590b815b-6e81-4d3d-98d3-54f08b9763d8) 

## Definition of Done

- [x] Add a new section, "Hidden tags," to the Dashboard:
  - [x] All Tags that the user follows with a tag weight < 0 (i.e., negative) will now only appear within the "Hidden Tags" section.
  - [x] Within the "Hidden tags" section, each tag will be displayed with an "Unhide" button, which will remove the tag from the user's tag follows entirely.
  - [x] No feed weights will be displayed on Tags in the "Hidden tags" section.
- [x] Additionally, only Tags with Tag Weights >= 0 (i.e., 0 and positive) will appear in the "Following tags" section.
  - [x] Within the "Following tags" section, each tag will be displayed with an "Unfollow" button, which users can press to stop following the tag.
  - [x] Additionally, each tag will have a "Hide" option hidden beneath a meatball/three-dot button, which, when pressed, will set a tag weight of -1 on the tag (and consequently move the tag to the "Hidden tags" section).
  - [x] No feed weights will be displayed on Tags in the "Following tags" section. Existing tag weights will remain in the database in case we want to roll back this change for any reason.
- [x] On the Tags page (`/tags`), each Tag will show an additional "Hide" toggle, which will "follow" the tag for the user and set the tag weight to -1 on the tag. This will show for both "Top Tags" and the tag search results.
  - [x] The tags page will have a slight re-design to update the existing view and better support the new functionality
  - [x] On the Tags page, a tag can either be:
    - [x] Default, unfollowed and unhidden, in which both a "Follow" and "Hide" button will show
    - [x] Followed, in which a "Following" toggle button will show, which, when pressed, will unfollow the tag.
    - [x] Hidden, in which an "Unhide" toggle button will show, which, when pressed, will unhide the tag.

**Comment by Ridhwana** [comment]
HI 👋 I just want to make a note on this issue that this feature is more complicated then anticipated (that box of tech debt @PhilipHow) because once we render the view, we immediately try to paginate it using this file called `initScrolling`. `initScrolling` much like the `stories_controller` handles the scrolling for multiple resources like `organizations`, `users`, `podcasts`, `tags` and it relies on the convention and a pattern for the resources. 

In theory "hidden tags" is a `following` and `tag` resource with negative points hence it should re-use the same resources as following tags. However, the way that initScrolling is written makes it difficult for that re-use because its only expecting one resource of that type.

This also touches both the `dashboard_controller` and the `following_controller` where the dashboard controller renders the view server side and then the following_controller hydrates it. I'm not sure if this was taken into account in the estimate but I feel like its going to increase the estimate.  

I'm working through it at the moment, and trying to figure out the best solution. If I get blocked at any point I'll reach out to the team for a second pair of eyes as this is a p1. At the same time I'm making notes on how we could possibly improve this by refactoring it out in a separate task so that it does not slow us down in the future. We have task on the board to migrate the `initScroling` file to a packs file, in which we should write tests and possibly refactor this out more. I've already made a note [there](https://github.com/forem/forem/issues/19709#issuecomment-1657276012) and added a forem team label because it touches super important parts of the application.  

Theres no action item here, but I wanted to give a progress update as I've done about 2-3 days of work on this thus far. I've separated out the PRs so that we can ship it in smaller parts and make incremental progress. 

cc @mirie @PhilipHow

**Comment by Ridhwana** [comment]
https://www.loom.com/share/2756264d7b424515b8988c4462a7958f?sid=b93e4914-03d2-44cc-a287-1e2c9b8bcb87

@PhilipHow sorry for the terrible audio :( let me know if you need me to re-record. I had a question regarding how we can potentially release this?

**Comment by PhilipHow** [comment]
Thank you for the update @Ridhwana. There is some really well-considered work here.

I'm happy with splitting up the work into multiple PRs as you've done/suggested and releasing incrementally.

**Comment by Ridhwana** [comment]
Dropping some information here that I hope use to create a knowledge sharing article for:

I'm working on adding the hide and unhide button on the tags page. The way this is currently built is that we render the initial page from the server with all of then unfollowed. Then through a javascript file we hydrate the view by determining whether the tag is followed or not. 

We do this by searching for all follow buttons that are not a user followed button (this includes organization follow buttons, podcast follow button, tag follow buttons etc.). We identify this by id's in the DOM. 

Then we loop over each button to fetch the correct status. For organizations and podcasts we do this by calling an endpoint `/follows/${buttonInfo.id}?followable_type=${buttonInfo.className}`. For tags, we do this by using the `cached_following_tags` on the user data-attribute on the `body` element in the DOM. To amend this for hidden tags, we need to separate these `cached_following_tags` by points and determine the hide/unhide button based on that.

**Comment by Ridhwana** [comment]
As I go through the code, I notice that on a signed out page, when a user clicks on "follow" and it opens up a modal for signup, we show a signup modal that we add a tracking event for. I'm assuming that we want the same for the "hide" button and we should add it as a follow up issue.

**Comment by Ridhwana** [comment]
@mirie and @PhilipHow, some updates on this: 

- I've released a change to the Tags page that account for some of the [UI changes](https://github.com/forem/forem/pull/19866). Changing the buttons styles are more complicated because they require us to move to using c-btn. Read more about `crayons-btn` and `c-btn` [here](https://forem-team.slack.com/archives/C7GE84DEJ/p1689246155762189). However, the changes will not be scoped to the tag page but will affect the home page, podcasts, organization page because the js code for the pagination is being used for all those resources. Hence, I'll tackle this last so that it does not hold back the release of the functionality.  There's also some other tweaks that need to be made. 

<img width="1512" alt="Screenshot 2023-08-10 at 12 59 30" src="https://github.com/forem/forem/assets/2786819/fbaadfac-a6d4-4ad1-a782-5d73c08372aa">

-  Today I will be deploying the change to [Update following_tags page and create hidden_tags page](https://github.com/forem/forem/pull/19873). This will account for removing the negative weights from dashboard/following_tags and adding them to dashboard/hidden_tags.

- I am currently working on showing the hide and follow button on the tags page which in theory sounds like a simple thing to implement but again I've described some of the [tradeoffs](https://github.com/forem/forem/issues/19675#issuecomment-1669341486) I've had to make. I think I could be done with this possibly by end of tomorrow and then I'll start working on the hide/follow on dashboard/following_tags and dashboard/hidden_tags.

- Thereafter, I'll focus on some UI polishing and cleanups, writing up the additional refactor tasks etc. 

I suspect I _could be_ completed with the main functionality before the end of next week. In addition, as mentioned earlier I will try to release iteratively to ensure that we see benefit earlier rather than later.

**Comment by Ridhwana** [comment]
I've been doing Slack updates with Phil and an update to Mai in our 1-1 + standups but I think for transparency and document purposes I'm overdue for an update on the issue. 

So in terms of functionality I've completed the dashboard pages (hidden and following) and shipped that feature last week. I had to log a [follow up bug](https://github.com/forem/forem/issues/19998) due to the caching. As discussed with Phil, I'll describe the technical challenges and proposal some improvements in our Snack n Study later this week. 

I am currently working with the /tags page and implementing the 'hide' functionality there. Within the [`followButtons.js` ](https://github.com/forem/forem/blob/b50077c7035d63a15163ccc2a82ffc338bcfd5bc/app/javascript/packs/followButtons.js) packs we tackle all Follow buttons in the same manner. However, by introducing the hide functionality the rules are now changed and its been challenging to retrofit "hide" into this file as its already very abstract. An initial (really ugly pass) of the code looked like [this](https://github.com/forem/forem/pull/19921) which honestly was adding a ton more technical debt, really difficult to follow and very brittle. If you look closely at the followButtons pack file you'll notice that the main purpose of the file is to change the state of the DOM using vanilla JavaScript. However, Preact seems like a better fit for this and would make the code more legible and would maintain the state for us so that we don't have to manually change the elements. 

Hence, I made the decision to rather implement the tags page hydration in Preact. The implementation is going much faster, its easier to read and feels like an overall improvement. I chatted to Joshua about this for a sanity check.  Whilst I'm only tackling the tags page, i think possibly all the follow buttons (i.e. the followButtons pack file) could benefit from a similar architecture. 

I'm _really really_ close to having this PR ready and feeling optimistic that it should be complete within the next day. 

Once this is complete this task is essentially functionally complete. We'd need to make some UI tweaks, wrap up any odds and ends, document my findings and maybe come up with a plan for tackling some of the debt.

Hope this update provides clarity!
